### PR TITLE
Fix type adaptation when there is loops

### LINKED ISSUE #86: todo: Typage adaptatif
Si ya bien une killer feature dans Nit c'est le typage adaptatif.

Le pb est la spécification du typage dans les cas de boucles.

Actuellement, la regle assez compliquée et assez limitative.
Le danger serait d'avoir une règle expressive mais trop complexe.

On peut mesurer la complexité de la règle par la taille de la doc dans la spec + la taille du message d'erreur

Ce que l'on voudrait c'est une règle simple et expressive.
On voudrait aussi un règle qui fait que les cas sont stables.

Actuellement, le prog suivant est refusé

    var x: A
    x = new B
    while ... do
       # zone 1
       x = new A # Erreur x est statiquement un B
    end

En effet, qui nous dit que dans la zone 1 on ne profite pas du fait que a est statiquement un B ?

Il faudrait refaire un tout pour vérifier.
On en toute généralité le nombre de tours semble etre infini puisque que le nombre de type statique est infini (TODO: trouver un exemple ou x passe par une infinité de types)

### LINKED ISSUE #647: typing: wrong behaviour with nullables
Here an example:

~~~
redef class Int
        fun next: nullable Int do if self < 20 then return self + 1 else return null
end

var t2: nullable Int = 10
while t2 != null do
        while t2 != null do
                # t2 = null # will infinite loop
                # t2 = t2.next # will seg fault
        end
end
~~~

Same issue with `nitg-s` and `nitg-g` but not with the interpretor.

### PR DESCRIPTION
**Pull request by privat**
Basically one of the oldest bug of Nit.

The point of adaptive typing is that the static types of variables follows the static types of the assigned values, type tests and comparison to null.
This is a great feature of the language but was buggy because loops where not taken in account (old TODO), until the present PR.

The basic idea is just to track if a type adaptation occurs, if yes replay the whole typing phase.
The immediate drawback is that now typing require 2 passes most of the time, and sometime more.

This PR was a roller-coaster of emotions to develop since a lot a surprise issues were discovered.

* bug in flow that give me hours of despair.
* duplication of error messages; because multiple passes.
* premature warning/error messages; because some error conditions on types are detected at the first pass but a second pass find it is fine in fact... but too late the error message was printed.
* almost work at the first time (modulo the previous issues)
* only two bugs found in the whole code, quite impressive, and one of them is not a real bug but an autocast issue.

The issue about error duplication and premature errors is only workadounded (or fixme) in most places.
But proper solutions will be required in future PR.

close #647 and close #86, then reopen it again, then reclose it.

### CODE REVIEW

**Comment by xymus** [comment]
:shipit: !

**Comment by privat** [comment]
reroll: fix fixmes of the previous roll about dead untypes code, ant try to pass tests

**Comment by R4PaSs** [comment]
:shipit: !

File: src/semantize/typing.nit [hunk 1]
```diff
@@ -540,49 +552,58 @@ end
 redef class Variable
 	# The declared type of the variable
 	var declared_type: nullable MType
+
+	# Was the variable type-adapted?
+	# This is used to speedup type retrieval while it remain `false`
```
Comment (Morriar) [inline src/semantize/typing.nit]: remains*

**Comment by Morriar** [comment]
+1

**Pull request merged**